File: agent/cmd/cuinterpose/rust/core/src/control.rs
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES.
// SPDX-License-Identifier: Apache-2.0

//! Two prestarted workers separate peer FD service from serialized CUDA control.
//! Queue pressure refuses requests before mutation; no operation is retried.

use super::process::Socket;
use super::state::{self, Result};
use cudarc::driver::sys::CUresult::CUDA_ERROR_NOT_INITIALIZED;
use cuinterpose_protocol::{self as protocol, NamespacePid, Operation, Reply, Request, Response};
use rustix::event::{PollFd, PollFlags, poll};
use std::os::unix::fs::PermissionsExt;
use std::os::unix::net::{UnixListener, UnixStream};
use std::sync::atomic::Ordering;
use std::sync::mpsc::{self, TrySendError};

// One running control operation and at most eight waiting connections. Peer
// exports never enter this queue: reciprocal importers need them to progress.
const CONTROL_QUEUE_CAPACITY: usize = 8;

enum ControlRequest {
    Inspect,
    Execute(Operation),
}

/// Private workers cannot dispatch until the single listener handoff succeeds.
/// Dropping this owner cancels them without joining: a caller may hold the
/// loader lock needed by a worker's Rust TLS startup or teardown.
pub struct PreparedWorkers {
    activation: mpsc::SyncSender<Socket<UnixListener>>,
    // Owned only between a successful bind and the worker handoff.
    listener: Option<Socket<UnixListener>>,
}

impl PreparedWorkers {
    pub fn prepare(namespace_pid: NamespacePid) -> Result<Option<Self>> {
        let (sender, receiver) =
            mpsc::sync_channel::<(Socket<UnixStream>, ControlRequest)>(CONTROL_QUEUE_CAPACITY);
        let (activation, parked) = mpsc::sync_channel::<Socket<UnixListener>>(1);
        let _worker = std::thread::Builder::new()
            .name("cuinterpose-control".into())
            .spawn(move || {
                // Queued sockets remain in the atfork FD registry.
                while let Ok((socket, request)) = receiver.recv() {
                    crate::boundary::call(&super::G_FAILED, (), || {
                        let _ = serve(socket, request, namespace_pid);
                    });
                }
            })
            .map_err(|error| {
                eprintln!("cuinterpose: control worker startup failed: {error}");
                CUDA_ERROR_NOT_INITIALIZED
            })?;
        if state::initialized() {
            return Ok(None);
        }
        let started = std::thread::Builder::new()
            .name("cuinterpose-peer".into())
            .spawn(move || {
                let Ok(listener) = parked.recv() else {
                    return;
                };
                loop {
                    let mut events = [PollFd::new(&*listener, PollFlags::IN)];
                    match poll(&mut events, None) {
                        Err(rustix::io::Errno::INTR) => continue,
                        Ok(_) if events[0].revents() == PollFlags::IN => {}
                        _ => {
                            super::G_FAILED.store(true, Ordering::Release);
                            break;
                        }
                    }
                    let Ok(socket) = Socket::open(|| listener.accept().map(|(socket, _)| socket))
                    else {
                        continue;
                    };
                    crate::boundary::call(&super::G_FAILED, (), || {
                        let _ = dispatch(socket, namespace_pid, &sender);
                    });
                }
            });
        if let Err(error) = started {
            // The failed closure drops the only control-queue sender.
            eprintln!("cuinterpose: peer listener startup failed: {error}");
            return Err(CUDA_ERROR_NOT_INITIALIZED.into());
        }
        Ok(Some(Self {
            activation,
            listener: None,
        }))
    }

    /// No spawn, blocking channel operation, formatting, or callback is allowed
    /// here. The caller holds the generation installation lock.
    /// With pinned Rust/glibc, mutexes and try_send wakeups use futexes and
    /// non-Drop TLS, not loader registration. The channel is preallocated;
    /// socket registration may grow its Vec using the ordinary glibc allocator.
    /// Eager ELF binding prevents first-use loader lookup in these libc calls.
    pub fn activate(&mut self, endpoint: &str) -> Result<()> {
        self.listener = Some(
            Socket::open(|| UnixListener::bind(endpoint))
                .map_err(|_| CUDA_ERROR_NOT_INITIALIZED)?,
        );
        let listener = self.listener.as_ref().unwrap();
        listener
            .set_nonblocking(true)
            .map_err(|_| CUDA_ERROR_NOT_INITIALIZED)?;
        std::fs::set_permissions(endpoint, std::fs::Permissions::from_mode(0o600))
            .map_err(|_| CUDA_ERROR_NOT_INITIALIZED)?;
        match self.activation.try_send(self.listener.take().unwrap()) {
            Ok(()) => Ok(()),
            Err(TrySendError::Full(listener) | TrySendError::Disconnected(listener)) => {
                self.listener = Some(listener);
                Err(CUDA_ERROR_NOT_INITIALIZED.into())
            }
        }
    }

    pub fn cleanup(&mut self, endpoint: &str) {
        // Only successfully bound, unpublished endpoints belong to this owner.
        // Cleanup is deliberately outside the installation lock.
        if let Some(listener) = self.listener.take() {
            drop(listener);
            let _ = std::fs::remove_file(endpoint);
        }
    }
}

fn dispatch(
    socket: Socket<UnixStream>,
    namespace_pid: NamespacePid,
    sender: &mpsc::SyncSender<(Socket<UnixStream>, ControlRequest)>,
) -> protocol::Result<()> {
    // Classification uses per-I/O socket timeouts, not a total header deadline.
    // A slow peer can delay acceptance, but never waits on STATE or lifecycle
    // CUDA calls. Fork during active protocol traffic is outside the contract.
    let timeout = Some(cuinterpose_protocol::timeout(None));
    socket.set_read_timeout(timeout)?;
    socket.set_write_timeout(timeout)?;
    let (request, descriptor): (Request, _) = protocol::receive(&socket)?;
    let addressed = match &request {
        Request::Inspect {
            namespace_pid: target,
        }
        | Request::Execute {
            namespace_pid: target,
            ..
        } => *target == namespace_pid,
        Request::Export { allocation } => allocation.creator_pid == namespace_pid,
    };
    if descriptor.is_some() || !addressed {
        return refuse(
            &socket,
            namespace_pid,
            "invalid cuinterpose control request",
        );
    }
    let request = match request {
        Request::Inspect { .. } => ControlRequest::Inspect,
        Request::Execute { operation, .. } => ControlRequest::Execute(operation),
        Request::Export { allocation } => {
            if super::G_FAILED.load(Ordering::Acquire) {
                return refuse(&socket, namespace_pid, "cuinterpose state failed");
            }
            let cache = match state::cache() {
                Ok(cache) => cache,
                Err(_) => {
                    return refuse(&socket, namespace_pid, "creator resource is unavailable");
                }
            };
            return cache.send(&socket, namespace_pid, &allocation.id);
        }
    };
    match sender.try_send((socket, request)) {
        Ok(()) => Ok(()),
        Err(error) => {
            let ((socket, _), message) = match error {
                TrySendError::Full(request) => {
                    (request, "control queue full; refused without mutation")
                }
                TrySendError::Disconnected(request) => (
                    request,
                    "control worker unavailable; refused without mutation",
                ),
            };
            refuse(&socket, namespace_pid, message)
        }
    }
}

fn refuse(socket: &UnixStream, namespace_pid: NamespacePid, message: &str) -> protocol::Result<()> {
    protocol::send(
        socket,
        &Response {
            namespace_pid,
            result: Err(message.into()),
        },
        None,
    )
}

fn serve(
    socket: Socket<UnixStream>,
    request: ControlRequest,
    namespace_pid: NamespacePid,
) -> protocol::Result<()> {
    let loading = matches!(request, ControlRequest::Execute(Operation::LoadAllocations));
    let result = (|| -> std::result::Result<Reply, String> {
        if super::G_FAILED.load(Ordering::Acquire) {
            return Err("cuinterpose state failed".into());
        }
        let mut state = state::get().map_err(|_| "cuinterpose state is unavailable")?;
        match request {
            ControlRequest::Inspect => {
                let live_raw_imports = state.live_raw_imports();
                let unsupported_creations = state.unsupported_exportable_creations();
                let records = state
                    .inspect()
                    .map_err(|_| "cannot inspect current CUDA state")?;
                Ok(Reply::Inspection {
                    records,
                    live_raw_imports,
                    unsupported_creations,
                })
            }
            ControlRequest::Execute(operation) => {
                state
                    .validate_lifecycle(operation)
                    .map_err(|_| "CUDA lifecycle operation refused without mutation")?;
                let result = if matches!(
                    operation,
                    Operation::RestoreMulticastCreators
                        | Operation::RestoreMulticastImporters
                        | Operation::RestoreMulticastDevices
                        | Operation::RestoreMulticastBindings
                ) {
                    super::multicast::restore_phase(state, operation)
                        .map(|bytes| super::host_carrier::Transfer { bytes, copy_us: 0 })
                } else {
                    state.lifecycle(operation)
                };
                match result {
                    Ok(transfer) => Ok(Reply::Completed {
                        operation,
                        bytes: transfer.bytes,
                        copy_us: transfer.copy_us,
                    }),
                    Err(code) => {
                        super::G_FAILED.store(true, Ordering::Release);
                        Err(format!("CUDA lifecycle operation failed: {code}"))
                    }
                }
            }
        }
    })();
    let loaded = loading && result.is_ok();
    protocol::send(
        &socket,
        &Response {
            namespace_pid,
            result,
        },
        None,
    )?;
    if loaded
        && let Ok(mut state) = state::get()
        && let Some(arena) = state.arena.take()
        && arena.release().is_err()
    {
        super::G_FAILED.store(true, Ordering::Release);
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn disconnected_activation_retains_listener_for_unlocked_cleanup() {
        let directory =
            std::env::temp_dir().join(format!("cuinterpose-activation-{}", std::process::id()));
        std::fs::create_dir(&directory).unwrap();
        let endpoint = directory.join("control.sock");
        let endpoint = endpoint.to_str().unwrap();
        let (activation, receiver) = mpsc::sync_channel(1);
        drop(receiver);
        let mut workers = PreparedWorkers {
            activation,
            listener: None,
        };
        assert!(workers.activate(endpoint).is_err());
        assert!(workers.listener.is_some());
        workers.cleanup(endpoint);
        assert!(!std::path::Path::new(endpoint).exists());
        assert!(workers.listener.is_none());
        std::fs::remove_dir(directory).unwrap();
    }
}

File: agent/cmd/cuinterpose/cpp/control.cpp
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES.
// SPDX-License-Identifier: Apache-2.0

#include "state.hpp"
#include <cstring>
#include <poll.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <sys/un.h>
#include <unistd.h>

namespace cuinterpose {
Fd request_export(const Ticket& ticket) try {
    ticket.validate();
    auto kind = std::holds_alternative<std::monostate>(ticket.resource) ? ResourceKind::Unicast : ResourceKind::Multicast;
    // Outgoing peer sockets belong to the same atfork registry as listeners
    // and queued connections, including while CUDA collective calls are unlocked.
    auto socket = open_socket([&] { return connect(ticket.endpoint, timeout()); });
    send(socket->get(), Request{Export{ticket.creator, kind, ticket.allocation}});
    auto received = receive(socket->get());
    auto response = received.body.get<Response>();
    if (std::holds_alternative<Failure>(response.result)) throw CudaError{invalid_handle};
    auto exported = std::get_if<Exported>(&std::get<Reply>(response.result));
    if (response.participant != ticket.creator || !exported || exported->resource != kind ||
        exported->allocation != ticket.allocation || received.descriptor.get() < 0)
        throw CudaError{invalid_handle};
    return std::move(received.descriptor);
} catch (const Json::exception&) {
    throw CudaError{invalid_handle};
} catch (const std::invalid_argument&) {
    throw CudaError{invalid_handle};
} catch (const std::runtime_error&) {
    // Refusal, a missing creator, and malformed/closed peer connections are
    // ordinary import failures. Allocation failures and unexpected exceptions
    // still reach the ABI failure boundary. During restore, the lifecycle
    // handler makes even this recoverable CUDA error fatal to reconstruction.
    throw CudaError{invalid_handle};
}
namespace {
struct Work { std::unique_ptr<Socket> socket; Request request; };
struct Queue {
    std::mutex mutex;
    std::condition_variable ready;
    std::deque<Work> work;
    bool stopped{};
};
void refuse(int socket, Id id, const std::string& message) {
    send(socket, Response{id, Failure{message}});
}
void serve(Work work, Generation& current) {
    std::variant<Reply, Failure> response;
    bool loaded = false;
    try {
        auto lock = lock_state(false);
        auto& state = current.state;
        if (std::holds_alternative<Handshake>(work.request)) response = Reply{Handshake{}};
        else if (std::holds_alternative<Inspect>(work.request)) {
            auto stats = state.stats();
            try {
                response = Reply{Inspection{state.inspect(), stats.live_raw_imports, stats.unsupported_exportable_creations}};
            } catch (const CudaError&) {
                response = Failure{"cannot inspect current CUDA state"};
            }
        } else {
            auto operation = std::get<Execute>(work.request).operation;
            state.validate_lifecycle(operation);
            try {
                Transfer transfer;
                if (operation >= Operation::RestoreMulticastCreators) restore_multicast(operation, lock);
                else transfer = state.lifecycle(operation);
                response = Reply{Completed{operation, transfer.bytes, transfer.copy_us}};
                loaded = operation == Operation::LoadAllocations;
            } catch (...) {
                failed.store(true);
                throw;
            }
        }
    } catch (const CudaError& error) { response = Failure{"CUDA lifecycle operation refused or failed: " + std::to_string(error.code)}; }
    catch (const std::exception& error) { failed.store(true); response = Failure{error.what()}; }
    try {
        send(work.socket->get(), Response{current.state.identity, std::move(response)});
    } catch (const std::system_error&) {
        // Losing a read-only client cannot invalidate CUDA state. A lifecycle
        // reply failure still reaches the worker's fail-closed boundary.
        if (std::holds_alternative<Execute>(work.request)) throw;
        return;
    }
    if (loaded) {
        auto lock = lock_state(false);
        auto arena = std::move(current.state.arena);
        if (arena && arena->release()) failed.store(true);
    }
}
}
void start_control(Generation& current) {
    auto listener = open_socket([] {
        int descriptor = ::socket(AF_UNIX, SOCK_STREAM | SOCK_CLOEXEC, 0);
        if (descriptor < 0) throw std::system_error(errno, std::generic_category(), "socket");
        return Fd(descriptor);
    });
    sockaddr_un address{};
    address.sun_family = AF_UNIX;
    std::memcpy(address.sun_path, current.state.endpoint.c_str(), current.state.endpoint.size()+1);
    if (::bind(listener->get(), reinterpret_cast<sockaddr*>(&address), sizeof(address)))
        throw std::system_error(errno, std::generic_category(), "bind control socket");
    auto queue = std::make_shared<Queue>();
    try {
        if (::chmod(current.state.endpoint.c_str(), 0600) || ::listen(listener->get(), 128))
            throw std::system_error(errno, std::generic_category(), "listen");
        std::thread worker([queue, &current] {
            try {
                for (;;) {
                    Work work;
                    {
                        std::unique_lock lock(queue->mutex);
                        queue->ready.wait(lock, [&] { return queue->stopped || !queue->work.empty(); });
                        if (queue->work.empty()) return;
                        work = std::move(queue->work.front());
                        queue->work.pop_front();
                    }
                    try { serve(std::move(work), current); }
                    catch (...) { failed.store(true); }
                }
            } catch (...) { failed.store(true); }
        });
        worker.detach();
        try {
            std::thread peer([queue, listener = std::move(listener), &current] {
                try {
                    for (;;) {
                        pollfd event{listener->get(), POLLIN, 0};
                        int result;
                        do { result = ::poll(&event, 1, -1); } while (result < 0 && errno == EINTR);
                        if (result != 1 || event.revents != POLLIN) { failed.store(true); return; }
                        auto socket = open_socket([&] {
                            int descriptor = ::accept4(listener->get(), nullptr, nullptr, SOCK_CLOEXEC);
                            if (descriptor < 0) throw std::system_error(errno, std::generic_category(), "accept");
                            return Fd(descriptor);
                        });
                        try {
                            timeval value{timeout().count(), 0};
                            if (::setsockopt(socket->get(), SOL_SOCKET, SO_RCVTIMEO, &value, sizeof(value)) ||
                                ::setsockopt(socket->get(), SOL_SOCKET, SO_SNDTIMEO, &value, sizeof(value)))
                                throw std::system_error(errno, std::generic_category(), "timeout");
                            auto received = receive(socket->get());
                            auto request = received.body.get<Request>();
                            bool identified = std::visit([&](const auto& item) {
                                if constexpr (requires { item.participant; }) return item.participant == current.state.identity;
                                else return true;
                            }, request);
                            if (received.descriptor.get() >= 0 || !identified) {
                                refuse(socket->get(), current.state.identity, "invalid cuinterpose control request");
                                continue;
                            }
                            if (auto export_request = std::get_if<Export>(&request)) {
                                if (failed.load()) { refuse(socket->get(), current.state.identity, "cuinterpose state failed"); continue; }
                                try {
                                    auto lease = current.cache.acquire({export_request->resource, export_request->allocation});
                                    send(socket->get(), Response{current.state.identity, Reply{Exported{export_request->resource, export_request->allocation}}},
                                         lease->descriptor.get());
                                } catch (...) { refuse(socket->get(), current.state.identity, "creator resource is unavailable"); }
                            } else {
                                std::lock_guard lock(queue->mutex);
                                if (queue->work.size() == 8) refuse(socket->get(), current.state.identity, "control queue full; refused without mutation");
                                else { queue->work.push_back({std::move(socket), std::move(request)}); queue->ready.notify_one(); }
                            }
                        } catch (...) { /* Malformed/closed connections do not poison CUDA state. */ }
                    }
                } catch (...) { failed.store(true); }
            });
            peer.detach();
        } catch (...) {
            std::lock_guard lock(queue->mutex);
            queue->stopped = true; queue->ready.notify_one();
            // Never join the worker while a constructor may hold loader locks.
            throw;
        }
    } catch (...) { ::unlink(current.state.endpoint.c_str()); throw; }
}
}
